### ISSUE TITLE
`setState` mutates state instead of replacing it

### ISSUE THREAD

**Issue by Mitranim** [developit/preact#1168]
Currently in Preact `8.2.9`, `Component.prototype.setState` mutates the previous state instead of replacing it with a patched version. This is unnecessarily inconsistent with React, can break user intuitions, can break things ported from React, and makes the second argument to `shouldComponentUpdate` useless and misleading.

I discovered this the hard way when sharing an initial state constant between components, assuming it would be treated as immutable. As previously mentioned, this can break user intuitions when coming from React.

Can we fix this? I would be willing to do a PR.

**Comment by developit** [comment]
Definitely definitely happy to see a PR for this if you have the time. In the meantime do you have a concise test we could use to reproduce the issue?

**Comment by Mitranim** [comment]
Here's a simple reproduction (run in Node):

```js
'use strict'

const preact = require('preact')
const render = require('preact-render-to-string')

const STATE = {}

class Component extends preact.Component {
  constructor() {
    super(...arguments)
    this.state = STATE
  }

  componentWillMount() {
    this.setState({key: 'value'})
  }
}

console.info(STATE)
// {}

render(preact.h(Component))

console.info(STATE)
// { key: 'value' }

// Expected behavior: STATE remains {}
```

**Comment by KevinDoughty** [comment]
Hello I've fixed this bug but have never contributed before. I'm willing to send a PR or describe the fix for someone else. I'm a bit hesitant since I'm new to the codebase. I have a test and two modifications.

```js
	it("issue 1168, state should not be mutated", () => {
		const STATE = {};
		class C1168 extends Component {
			constructor() {
				super(...arguments);
				this.state = STATE;
			}
			componentWillMount() {
				this.setState({key: 'value'});
			}
		}
		render(<C1168 />, scratch);
		expect(STATE).to.deep.equal({});
	});
```

src/component.js Component.prototype:
```js
	setState(state, callback) {
		if (!this.prevState) this.prevState = this.state;
		this.state = extend(extend({},this.state), typeof state==='function' ? state(this.state, this.props) : state);
		if (callback) this._renderCallbacks.push(callback);
		enqueueRender(this);
	},
```

src/vdom/component.js renderComponent:
```js
	if (component.constructor.getDerivedStateFromProps) {
		state = extend({},state);
		component.state = extend(state, component.constructor.getDerivedStateFromProps(props, state));
	}
```

Bundle size decreases by three bytes, from 3494 to 3491, and there is not a noticeable performance difference. I did not, however, check if this has an effect on issue 1169, but will shortly.

**Comment by KevinDoughty** [comment]
Oh I guess someone beat me to it, with issue 1170.

**Comment by developit** [comment]
This should be fixed in [8.3.0](https://github.com/developit/preact/releases/tag/8.3.0)!
Please feel free to re-open otherwise :)
